A Sender MUST obey it's Media Profile (#12)
AMWA-TV/nmos-sink-metadata-processing: docs/1.2. Behaviour.md

## Senders

> The `/media-profiles` endpoint allows the client to pass an array of [Media Profiles](1.0.%20Overview.md#media-profile) to a Sender and let the underlying component to choose exactly which format from the set it will use. A client can pass a single Media Profile to specify the exact format.

Reviewer: ```suggestion
The `/media-profiles` endpoint allows the client to pass an array of [Media Profiles](1.0.%20Overview.md#media-profile) to a Sender and let the underlying implementation to choose exactly which format from the set it will use. A client can pass a single Media Profile to specify the exact format.
```

> The initial state of Media Profiles of a Sender MUST be an empty JSON array `[]`. This state indicated a Sender is unbound by an Media Profile. Creating a connection with such a Sender via [IS-05][IS-05] is not impacted by the Sink Metadata Processing API.

Reviewer: ```suggestion
The initial state of Media Profiles of a Sender MUST be an empty JSON array `[]`. This state indicated a Sender is unbound by a Media Profile. Creating a connection with such a Sender via [IS-05][IS-05] is not impacted by the Sink Metadata Processing API.
```

Once a Media Profile set is accepted, a Sender MUST transmit a Flow and Source which meets at least one profile. This constraint MUST be adhered to when an external change (i.e. not from the `/media-profiles` endpoint) modifies any aspect of the stream. If the change does not satisfy its Media Profiles, the Sender MUST NOT transmit the signal over the network. The Sender is REQUIRED to take the following actions:
- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).

> If a Sender, its associated Flow, and/or Source do not meet the restrictions of a the Media Profiles, the Sender MUST refuse an [IS-05][IS-05] `/active` requests which tries to set `master_enable` property to `true` by responding with a `409 Conflict` response.

Reviewer: ```suggestion
If a Sender, its associated Flow, and/or Source do not meet the restrictions of the Media Profiles, the Sender MUST refuse an [IS-05][IS-05] `/staged` requests which tries to set `master_enable` property to `true` by responding with a `409 Conflict` response.
```

## Senders

The `/media-profiles` endpoint allows the client to pass an array of [Media Profiles](1.0.%20Overview.md#media-profile) to a Sender and let the underlying implementation to choose exactly which format from the set it will use. A client can pass a single Media Profile to specify the exact format.

### Role of Media Profiles

The initial state of Media Profiles of a Sender MUST be an empty JSON array `[]`. This state indicated a Sender is unbound by a Media Profile. Creating a connection with such a Sender via [IS-05][IS-05] is not impacted by the Sink Metadata Processing API.

> Once a Media Profile set is accepted, a Sender MUST transmit a Flow and Source which meets at least one profile. This constraint MUST be adhered to when an external change (i.e. not from the `/media-profiles` endpoint) modifies any aspect of the stream. If the change does not satisfy its Media Profiles, the Sender MUST NOT transmit the signal over the network. The Sender is REQUIRED to take the following actions:

Author: ```suggestion
Once a Media Profile set is accepted, a Sender MUST have a Flow and Source which meets at least one profile. This constraint MUST be adhered to when an external change (i.e. not from the `/media-profiles` endpoint) modifies any aspect of the stream. If the change does not satisfy its Media Profiles, the Sender MUST NOT transmit the signal over the network. The Sender is REQUIRED to take the following actions:
```

- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).

> - MUST respond with `204 No Content` when answering `GET /media-profiles` to indicate it is no longer following or adhering to a set. This response SHOULD presist until the next valid `PUT /media-profiles` request.

Author: ```suggestion
- MUST respond with `205 Reset Content` when answering `GET /media-profiles` to indicate it is no longer following or adhering to a set. This response SHOULD persist until the next valid `PUT /media-profiles` request.
```

The `/media-profiles` endpoint allows the client to pass an array of [Media Profiles](1.0.%20Overview.md#media-profile) to a Sender and let the underlying implementation to choose exactly which format from the set it will use. A client can pass a single Media Profile to specify the exact format.

### Role of Media Profiles

The initial state of Media Profiles of a Sender MUST be an empty JSON array `[]`. This state indicated a Sender is unbound by a Media Profile. Creating a connection with such a Sender via [IS-05][IS-05] is not impacted by the Sink Metadata Processing API.

Once a Media Profile set is accepted, a Sender MUST transmit a Flow and Source which meets at least one profile. This constraint MUST be adhered to when an external change (i.e. not from the `/media-profiles` endpoint) modifies any aspect of the stream. If the change does not satisfy its Media Profiles, the Sender MUST NOT transmit the signal over the network. The Sender is REQUIRED to take the following actions:

- MUST set the [IS-05][IS-05] `/active`'s `master_enable` property to `false`.
- MUST update the [IS-04][IS-04] `subscription`'s `active` property to `false` (regardless of unicast or multicast).

> - MUST respond with `204 No Content` when answering `GET /media-profiles` to indicate it is no longer following or adhering to a set. This response SHOULD presist until the next valid `PUT /media-profiles` request.

Author: ```suggestion
- MUST respond with `205 Reset Content` when answering `GET /media-profiles` to indicate it is no longer following or adhering to a set. This response SHOULD persist until the next valid `PUT /media-profiles` request.
```

> When a Sender is in this violated state, responding with a `204 No Content`, This state indicated a Sender is unbound by a Media Profile and is similar to the initial state.

Author: ```suggestion
When a Sender is in this state, ie. responding with a `205 Reset Content`, this indicates a Sender is unbound by a Media Profile and is similar to the initial state.
```

The initial state of Media Profiles of any Sender is empty. Creating a connection with such a Sender via IS-05 doesn't involve the Sink Metadata Processing API.

- A `GET` request returns the last successfully applied Media Profiles array.
- A `PUT` request MUST be validated and if the Device supports the proposed media profile, it initiates an attempt to reconfigure the Device and update corresponding Flow, Source, and/or Sender properties. If successful, the operation MUST return the accepted Media Profiles array, otherwise an error MUST be returned and nothing changed.
- A `DELETE` request MUST clear the last successfully applied Media Profiles array.

When the `/media-profiles` structure changes as a consequence of PUT or DELETE (request is validated and accepted by the API server), then the associated IS-04 Sender MUST have its version updated (in registered mode this MUST update the registered resource).

If the new `/media-profiles` structure causes underlying IS-04 flows and sources to change, then their version MUST be updated (in registered mode this MUST update the registered resources).

> When an external change (i.e. not from the `/media-profiles` endpoint) modifies the associated Source and/or Flow, the Sender MUST respect its
> [Media Profiles](1.0.%20Overview.md#media-profile). If the change does not satisfy its Media Profiles the Sender MUST NOT transmit the signal over the network.

Reviewer: ```suggestion
When an external change (i.e. not from the `/media-profiles` endpoint) modifies a Source and/or a Flow of a Sender and/or the Sender itself, it MUST respect [Media Profiles](1.0.%20Overview.md#media-profile) of this Sender. If the change does not satisfy its Media Profiles the Sender MUST NOT transmit the signal over the network.
```

Author: Maybe just "any change" not simply those covered currently?

```suggestion
When an external change (i.e. not from the `/media-profiles` endpoint) modifies any aspect of the stream then the Sender MUST respect its
[Media Profiles](1.0.%20Overview.md#media-profile). If the change does not satisfy its Media Profiles the Sender MUST NOT transmit the signal over the network.
```

Reviewer: Sounds appropriate. The only thing which seems possibly wrong is that "Sender MUST respect Media Profiles". We can say the change must respect them or use passive voice.

The Sender is REQUIRED to take the following actions:

> - MUST set the IS-05 `/active`'s `master_enabled` property to `false`.

Reviewer: ```suggestion
- MUST set the IS-05 `/active`'s `master_enable` property to `false`.
```
